pynxtools/dataconverter/readers/multi/reader.py

```diff
@@ -0,0 +1,208 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""An example reader implementation for the DataConverter."""
+import os
+from dataclasses import dataclass, field
+from functools import partial
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from pynxtools.dataconverter.readers.base.reader import BaseReader
+from pynxtools.dataconverter.readers.utils import parse_flatten_json, parse_yml
+from pynxtools.dataconverter.template import Template
+
+
+def fill_wildcard_data_indices(config_file_dict, dims):
+    """
+    Replaces the wildcard data indices (*) with the respective dimension entries.
+    """
+
+    def get_vals_on_same_level(key):
+        key = key.rsplit("/", 1)[0]
+        for k, v in config_file_dict.items():
+            if k.startswith(key):
+                yield v
+
+    for key, value in list(config_file_dict.items()):
+        for dim in dims:
+            new_key = key.replace("*", dim)
+            new_val = value.replace("*", dim)
+
+            if (
+                new_key not in config_file_dict
+                and new_val not in get_vals_on_same_level(new_key)
+            ):
+                config_file_dict[new_key] = new_val
+
+        config_file_dict.pop(key)
+
+
+@dataclass
+class ParseJsonCallbacks:
+    """
+    Callbacks for dealing special keys in the json file.
+
+    Args:
+        attrs_callback (Callable[[str], Any]):
+            The callback to retrieve attributes under the specified key.
+        data_callback (Callable[[str], Any]):
+            The callback to retrieve the data under the specified key.
+        dims (List[str]):
+            The dimension labels of the data. Defaults to None.
+    """
+
+    special_key_map = {
+        "@attrs": lambda _, v: ParseJsonCallbacks.attrs_callback(v),
+        "@link": lambda _, v: ParseJsonCallbacks.link_callback(v),
+        "@data": lambda _, v: ParseJsonCallbacks.data_callback(v),
+    }
+    attrs_callback: Callable[[str], Any] = lambda v: v
+    data_callback: Callable[[str], Any] = lambda v: v
+    link_callback: Callable[[str], Any] = lambda v: {"link": v}
+    dims: List[str] = field(default_factory=list)
+
+    def apply_special_key(self, precursor, key, value):
+        """
+        Apply the special key to the value.
+        """
+        ParseJsonCallbacks.special_key_map.get(precursor, lambda _, v: v)(key, value)
+
+
+def parse_json_config(
+    file_path: str,
+    callbacks: Optional[ParseJsonCallbacks] = None,
+) -> dict:
+    """
+    Parses a json file and returns the data as a dictionary.
+    """
+    if callbacks is None:
+        # Use default callbacks if none are explicitly provided
+        callbacks = ParseJsonCallbacks()
+
+    config_file_dict = parse_flatten_json(file_path)
+    fill_wildcard_data_indices(config_file_dict, callbacks.dims)
+
+    for key, value in config_file_dict.items():
+        if not isinstance(value, str) or ":" not in value:
```

Comment:
I was wondering if I may be possible to give a list of options in the config file. Something like:
- first check in @attrs
- if the key doesn't exist in @attrs, check the eln
- if it's not there either, write a default or nothing at all

This would probably be achieved by passing a list in the config.

I guess this comes down to how we see the config file. I have been using it in a way that there are default configs for each subreader and that the user ideally doesn't have to change the config file when converting. So then such a list of options would allow to pass an ELN file where a field is missing without breaking the reader, but still have a default config file. This may also come in handy if you have an additional reader for metadata that only comes into place when you upload a metadata file as well (see what I started [here](https://github.com/FAIRmat-NFDI/pynxtools-xps/blob/c0788e5a8ba6e8c769d9ca956aded28d83dae512/pynxtools_xps/reader.py#L295)).

Reply:
Yes, I think this also relates to my thoughts to your comment below. Let's discuss it.